### ISSUE TITLE
opencaching.de / OKAPI: upload fieldnotes

### ISSUE THREAD

**Issue by hxdimpf** [cgeo/cgeo#14908]
### Tell us your idea!

for a quick introduction see https://github.com/cgeo/cgeo/issues/14907

In addition to the referenced function, I have also worked a PR for  uploading fieldnotes as I have seen there was a request for this almost 3 years ago by @Lineflyer. My PR is https://github.com/opencaching/okapi/pull/630

I have some difficulties here with the character encoding used by cgeo and before I go to deep, is cgeo still interested in consuming such a service?

Ideally I would want to receive the fieldnotes in `utf-8` encoding, I know (by experience) that cgeo uses an UTF-16LE encoding with no BOM. I have not been able to handle this format in PHP (OKAPI uses PHP).

You could keep a hybrid file, e.g one that has GC, OC, OP, and whatever other logs in it. would be filtering out OC logs and ignore everything else, however, cgeo when it writes an offline log for an OC cache, it must actually use the log types that OCDE understands. For instance a "Write Note" for GC is a "Comment" for OC. Look at the PR, there you see what log types the platform understands. From my practical use of cgeo in the field I know that it writes offline logs for OC cacheds but it uses in the UI and in  the file the log types from GC.

**Comment by SammysHP** [comment]
The uncommon encoding has historic reasons. It was specified/implemented that way a long, long time ago together with Garmin GPS receivers and hasn't changed since then.

**Comment by ztNFny** [comment]
> Do you know the concept of offline logging? If yes, then you know that you can "upload" fieldnotes to GC but you cannot upload fieldnotes to OC. Here is why: Because the folks who wrote the OC functionality in cgeo did not have an API method. I am here to finally provide cgeo with that year long missing function.

Yes, all of that is clear - no idea at all what you want to say with this comment. I was addressing these points from your post in my https://github.com/cgeo/cgeo/issues/14908#issuecomment-1828347218:

> You could keep a hybrid file, e.g one that has GC, OC, OP, and whatever other logs in it.

c:geo doesn't keep a file for notes, it's generated on export.

> For instance a "Write Note" for GC is a "Comment" for OC. Look at the PR, there you see what log types the platform understands. 

c:geo has looong implemented all the OKAPI log types, see https://github.com/cgeo/cgeo/blob/b220fd190711e751d10d056f0b01f476e3a05499/main/src/main/java/cgeo/geocaching/log/LogType.java#L23. Without that online logging wouldn't be possible. So again, is there a specific reason you're bringing that up? Or does your API use newly made up log types that don't match the normal OKAPI logtypes?

**Comment by hxdimpf** [comment]
I am not interested in online logging because obviously that works.

This whole thread is about offline logging and dealing with the uploading of fieldnotes to opencaching.de during the "Export Fieldnotes" Dialog in cgeo.

**Comment by hxdimpf** [comment]
> > Do you know the concept of offline logging? If yes, then you know that you can "upload" fieldnotes to GC but you cannot upload fieldnotes to OC. Here is why: Because the folks who wrote the OC functionality in cgeo did not have an API method. I am here to finally provide cgeo with that year long missing function.
> 
> Yes, all of that is clear - no idea at all what you want to say with this comment. I was addressing these points from your post in my [#14908 (comment)](https://github.com/cgeo/cgeo/issues/14908#issuecomment-1828347218):
> 
> > You could keep a hybrid file, e.g one that has GC, OC, OP, and whatever other logs in it.
> 
> "There is not "fieldnote file" stored in c:geo", it's generated on export.
> 
> > For instance a "Write Note" for GC is a "Comment" for OC. Look at the PR, there you see what log types the platform understands.
> 
> c:geo has looong implemented all the OKAPI log types, see
> 
> https://github.com/cgeo/cgeo/blob/b220fd190711e751d10d056f0b01f476e3a05499/main/src/main/java/cgeo/geocaching/log/LogType.java#L23
> 
> . Without that online logging wouldn't be possible. So again, is there a specific reason you're bringing that up? Or does your API use newly made up log types that don't match the normal OKAPI logtypes?

Here is a fieldnotes file, generated by cgeo today:

```
OCD3AD,2023-11-27T08:27:57Z,Write note,"Thx to Borstelmann  for 'Lindenallee',
```
Do you see the "Write note"? Do you understand that this is not an opencaching log type? It should read "Comment"

What you are showing me above from your code may well work for online logging but it does not work in fieldnotes.

Thx much for your time

**Comment by hxdimpf** [comment]
@ztNFny : And one more thing: Rather than throwing darts at me, please dicuss that in your team. I am almost certain that this is all a big misunderstanding on you side. There are folks on your team who could explain it to you for sure. even if I can't.

If you jointly come to the conclusion that all works nicely since a long time and for the eternal future, I won't ever bother you again and discard my PR. Again, I was just jumping in to help. Of course, as an openaching.de cacher, I would have found this functionality very very useful but I can live without it for sure.

**Comment by hxdimpf** [comment]
Ok, so if I read your comments wrong, then I suggest we reset the conversation. agree?

Forget everything I said about where fieldnotes may be stored on cgeo or how they may be generated. I don't kknow any of this and it is not important in the context of this issue.

We have already established clarity with respect to the log types that should show up in the fieldnotes for OC records, right? This is of course not only in the fieldnotes, it is also in the UI when the user clicks those log types and it is not limited to just "Write note" .... I think you know understand that point and there is no need to discuss this any further, you agree?

Back to the new OKAPI Service: Do you understand what this proposed function in okapi is meant for? If not, pls get back and I try my best to explain it.

**Comment by ztNFny** [comment]
> I am almost certain we again discuss different things here. Do you know whch dialog I mean?

Yes I know - and no, we're talking about the same thing - which is off-topic to this issue btw.

> also keep in mind that someone might never have GC active, always only uses OC and now he logs an OC cache and the UI offers to click on "Write note"?

The share of users using only OC is tiny. Most OC users are also using GC. Calling the same log type by different names inside c:geo would be very confusing.

You're of course free to open a request to have that changed but seeing that it was deliberately implemented this way, has been like this for years and is the sensible solution for the biggest share of OC users I don't think the team will agree.

**Comment by ztNFny** [comment]
> I have some difficulties here with the character encoding used by cgeo

Your service should definitly support UTF-16LE as that's "the standard" format. Standard being defined by the old Garmin implementation, most apps use that encoding when exporting fieldnotes.txt. So if you want your endpoint to be universally usable it must support UTF-16LE.

**Comment by eddiemuc** [comment]
I opened opencaching.de and quickly found a page where one can upload field notes: https://www.opencaching.de/field-notes/

I also found this (German) documentation which claims that c:geo field notes are supported: https://wiki.opencaching.de/index.php?title=Field_Notes

This confused me. It seems like field notes from c:geo are already supported at least by oc.de. Is this not working (any more), or is this already the change to okapi we see in place here?

Sore if this is a stupid question. I am not a regular oc user and never used field notes up to now (neither for GC nor for oc).

**Comment by hxdimpf** [comment]
Let is differentiate website from OKAPI

**Website**: had this support long time ago but it was also broken for years, since cgeo changed its encoding. Only recently a PR was made (and merged) to get it back going. I believe that PR didn't make it into the production website yet, but I will clarify. I am in contact with the person who did that PR to find out whether he has successfully tested cgeo's UTF-16LE.

OKAPI: a service never existed for this function @Lineflyer asked for it early 2021. Nothing happend. I jumped in and wanted to do it. My PR is in review bu one of the original OKAPI developers (Wojtek). Whether and when it will be merged is unclear and it is even more unclear how long it will take from then on until it hits the production API server.

And to be very explicit: It is not about importing a fieldnotes file through a browser, it is about cgeo sending a datastrem over a new API.

**Comment by ztNFny** [comment]
Regarding this whole issue: As the feature is not available in OKAPI nothing can be done in c:geo as of now. So my proposal would be to pause/close this issue for now - and then once it's actually implemented in OKAPI AND available to use in one of the OCs we can look into an implementation.

Or is there feedback you require?  If so that probably should be given in https://github.com/opencaching/okapi/pull/630

**Comment by hxdimpf** [comment]
Relating to the comment above. Yes, the feature is not (yet) available and no, we should not pause/close the issue. Instead and hereby I would like to offer an opportunity to the cgeo team to jointly review the definition of the Webservice in question. Please take some time and review the attached documentation of the service. Feel free to post questions or concerns if it isn't what the client application may expect. Please note, there is no mentioning of character encoding as I believe I have coded it now such that it just accepts whatever the client application may send. If it works to your liking, please also take the time to give me an OK such that we can proceed with integrating and publishing it as quick as the process may be cranked.

![Screenshot from 2023-12-01 10-16-51](https://github.com/cgeo/cgeo/assets/60329000/4d7167b8-7340-48fe-aa46-7458ab20d3e1)

**Comment by fm-sys** [comment]
Hi @hxdimpf, thanks for your work! 

Have you tested what happens when uploading a field note to GC which contains log types they do not know? Also can physical Garmin devices create field notes with the different log type names? If one of this is not true, it would be better to wrap it server side.

**Comment by hxdimpf** [comment]
Again we have to carefully establish a precise context. What is "uploading" exactly? There are three paths:

A) client app -> API service -> target platform (draft logs, no immediate commit)
B) client app -> device's file system ->  webbrowser -> target platform
C) Online logging: client app -> API service -> target platform (immediate commit)

Let me start with A)
Yes, I have tested this today in the light of the new webservice I am providing for OCDE. So far I lived under the assumption that the webservice provided by GC accepts "foreign records" and what I mean with this is records that do not start with "GC". Essentially I am talking about a filtering mechanism but this doesn't seem to work as one might expect.

It depends on the order of draft logs how GC behaves and pls note that the fieldnotes object is internally sorted by log date (as it appears to me at least). LIFO ... Last in first out. This means the log with the newest log date comes first in the fieldnotes object and it is the first log GC is looking at. Now if this log happens to be an OC log, GC terminatesthe parsing and rejects the data, all of it. If the logs are reversed, which means a GC log is the first in the file and the second is an OC log, it processes the GC log and silently ignores the OC log. Interestingly it does not give up parsing. If there is a 3rd log, which happens to be a GC log, it takes it and accepts it.

I'd consider it a bug in GC's implementation in that API service mehod. Either you do the filtering or you don't and if you do, the order in which those logs come in should not make a difference.

Now B)
This is where the Garmin question comes into play. This path is used by clients which cannot communicate via REST API services and this path is currently the only path that could ever be used in the context of OCDE as no web service existed.

This path, once worked but was broken for a long time, howevr it has been fixed in September but perhaps not published to production.

To my knowledge, all clients that ever generated a data object which we refer to as "fieldnotes" have made their design point to accommodate the GC platform definition. This means almost independent from the cache type and independent from who owns the cache, at their UI they present a selection of one out of 3 Options to the User: Found It, Write note, Didn't find it.

I believe but haven't checked  that cgeo, in addition to those three, recognizes a cache type "Event" and presents an "Attended" instead of "Found it". Some of this can indeed be "fixed at the receiving side" but I'd consider this a circumvention at best but never a real solution. Reason being the user when he selects a log type at his UI, can't really log what ultimately shows up on the target platform. For instance, take a webcam, if the client app logs a "Found it", the server side onf the platform cann of course detect the cache type and silently convert this into a "Webcam photo taken" (just an example). There won't ever be an option for a user to log a "Will attend" and even if that existed, today's client apps would not check whether the event date is in the past or in the future and in case it is in the past they must not allow a "Will attend". All of these intricate details are implemented for GC but not for OC.

Ideally a client application would want to pursue one unified path per platform, for online logging (C) as well as offline logging (A, B). I am not focusing on C here as it is obvious if you have understood A and B.

I am speaking Utopia here, perhaps it can't be made that way for pragmatic reasons (no development resources, priorities, OC being a niche only).

Ideally the client application should treat all target platforms as first class citizen. This is not the case today for any client apps I know. For instance, cgeo, it "sort of" handles OC (and others) but it doesn't in a unified way adapting to the target platform's capabilities, at least not with respect to path (A, B). Subsequently I am addressing cgeo:

It differeentiates GC and OC in at least two major aspects. (1) it renders platform specific map icons and (2) it targets the platform via a platform specific API (be it webservice or scraping).

What it should to for logging ist the following:

Recognize the target platform for the activity at hand, if it is an OC object, learn the following:

By cache type, by owned | not owned| by event data already passed or not passed, by current state of the cache (enabled / disabled): Figure out what log type are possible, and use those or better the log type's agreed upon names when presenting the user with an options dialog. Also, when putting the log type into a field notes object. Also when directly commiting the log (C)

For instance, if a cache is not owned by the current user, don't show a "Temporary Disable" log type .... plus all the rest of the full state machine, I am sure you got the point. For GC you do all this in great detail, but not for OC, unless I missed it (I am rather new on OC)

Now in the near future, when my OKAPI services are merged and integrated, there is a service that exists today but will be extended as I will show below:

`::services/apisrv/installation` which tells you the relevant details a client application need to know about a plaform's capabilities:

```
{
  "site_url": "https://opencaching.de/",
  "okapi_base_url": "https://opencaching.de/okapi/",
  "okapi_base_urls": [
    "https://opencaching.de/okapi/"
  ],
  "site_name": "PRODUCTION",
  "okapi_version_number": 1926,
  "okapi_revision": 1926,
  "okapi_git_revision": "dbed1d0d908dcc00a4aa7387de2e08b727dccee8",
  "registration_url": "https://opencaching.de/register.php",
  "mobile_registration_url": null,
  "image_max_upload_size": 15360000,
  "image_rcmd_max_pixels": 786432,
  "has_image_positions": true,
  "has_ratings": false,
  "geocache_passwd_max_length": 20,
  "has_draft_logs": true,
  "has_lists": true,
  "cache_types": [
    "Drive-in Cache",
    "Event Cache",
    "Math/Physics Cache",
    "Moving Cache",
    "Multicache",
    "Quiz Cache",
    "Traditional Cache",
    "Unknown cache type",
    "Virtual Cache",
    "Webcam Cache"
  ],
  "log_types": [
    "Archived",
    "Attended",
    "Available",
    "Found",
    "Locked",
    "Locked, invisible",
    "Not found",
    "Note",
    "Temporarily not available",
    "Will attend"
  ]
}
```

The two arrays at the bottom, `cache_types` and `log_types` give you the relevant piece of information to construct the logging state machine. If you tell me that you want to turn the OC target platform into a "1st class citizen" I can construct the fully populated datastructure over all cache_type, cache_state, cache_ownership an what log types are possible. The data you see above, still requires some logic to be implemented in code rather than just doing a table lookup.

P.S: You could use the `cache_type` to provide better suited iconson the map.

And last not least: we can talk! If you want this webservice to provide any aditional information about the platform I will make the above mentioned webservice to provide it for you,

**Comment by bekuno** [comment]
@hxdimpf
Is your PR at OC/OKAPI still valid or in work?

**Comment by hxdimpf** [comment]
> [@hxdimpf](https://github.com/hxdimpf) Is your PR at OC/OKAPI still valid or in work?

see here: https://github.com/cgeo/cgeo/issues/14907#issuecomment-3198384774
